File: src/nxapi.cpp
#include "nso_album_sync/nxapi.hpp"

#include "nso_album_sync/json.hpp"
#include "nso_album_sync/util.hpp"

#include <algorithm>
#include <chrono>
#include <ctime>
#include <fstream>
#include <iomanip>
#include <iterator>
#include <map>
#include <sstream>
#include <stdexcept>

#ifdef _WIN32
#include "nso_album_sync/windows_compat.hpp"
#else
#include <sys/stat.h>
#endif

namespace nso {
namespace {

constexpr char kZncaBaseUrl[] =
    "https://nxapi-znca-api.fancy.org.uk/api/znca";
constexpr char kNxapiAuthUrl[] =
    "https://nxapi-auth.fancy.org.uk/api/oauth/token";
constexpr char kUserAgent[] =
    "nso-album-sync/2.0.0 (+https://github.com/Dycool/NSO-Album-Sync)";
constexpr char kZncaClientVersion[] = "w8zSLBsxR7rVoGJA";
constexpr char kCoralLoginUrl[] =
    "https://api-lp1.znc.srv.nintendo.net/v4/Account/Login";

constexpr auto kVersionCacheLifetime = std::chrono::hours(6);
constexpr auto kVersionRetryLifetime = std::chrono::minutes(15);
constexpr auto kDefaultRateLimitBackoff = std::chrono::minutes(15);

std::vector<std::string> znca_headers(const std::string& nso_version) {
    return {
        "X-znca-Platform: Android",
        "X-znca-Version: " + nso_version,
        std::string("X-znca-Client-Version: ") + kZncaClientVersion,
    };
}

std::int64_t epoch_seconds(std::chrono::system_clock::time_point value) {
    return std::chrono::duration_cast<std::chrono::seconds>(
               value.time_since_epoch())
        .count();
}

std::chrono::system_clock::time_point time_from_epoch(std::int64_t value) {
    return std::chrono::system_clock::time_point(std::chrono::seconds(value));
}

void replace_cache_file(
    const std::filesystem::path& temporary,
    const std::filesystem::path& target) {
#ifdef _WIN32
    if (!MoveFileExW(
            temporary.c_str(), target.c_str(),
            MOVEFILE_REPLACE_EXISTING | MOVEFILE_WRITE_THROUGH)) {
        std::error_code ignored;
        std::filesystem::remove(temporary, ignored);
    }
#else
    std::error_code error;
    std::filesystem::rename(temporary, target, error);
    if (error) std::filesystem::remove(temporary, error);
#endif
}

std::chrono::seconds retry_after_delay(const std::string& value) {
    try {
        const auto seconds = std::stoll(value);
        return std::chrono::seconds(std::max<std::int64_t>(1, seconds));
    } catch (...) {
    }

    std::tm parsed{};
    std::istringstream input(value);
    input.imbue(std::locale::classic());
    input >> std::get_time(&parsed, "%a, %d %b %Y %H:%M:%S GMT");
    if (input.fail()) {
        return std::chrono::duration_cast<std::chrono::seconds>(
            kDefaultRateLimitBackoff);
    }

#ifdef _WIN32
    const std::time_t retry_at = _mkgmtime(&parsed);
#else
    const std::time_t retry_at = timegm(&parsed);
#endif
    if (retry_at <= 0) {
        return std::chrono::duration_cast<std::chrono::seconds>(
            kDefaultRateLimitBackoff);
    }

    const auto now = std::time(nullptr);
    return std::chrono::seconds(std::max<std::int64_t>(
        1,
        static_cast<std::int64_t>(retry_at - now)));
}

}  // namespace

NxapiClient::NxapiClient(
    HttpClient& http,
    std::string client_id,
    std::filesystem::path cache_file)
    : http_(http),
      client_id_(std::move(client_id)),
      cache_file_(std::move(cache_file)) {
    load_cache();
}

void NxapiClient::load_cache() {
    if (cache_file_.empty() || !std::filesystem::exists(cache_file_)) return;
    try {
        std::ifstream file(cache_file_, std::ios::binary);
        const std::string contents{
            std::istreambuf_iterator<char>(file),
            std::istreambuf_iterator<char>()};
        const auto json = Json::parse(contents);
        version_ = json.string("nsoVersion");
        version_expiry_ = time_from_epoch(json.integer("nsoVersionExpiresAt"));
        rate_limit_until_ = time_from_epoch(json.integer("rateLimitUntil"));
    } catch (...) {
        version_.clear();
        version_expiry_ = {};
        rate_limit_until_ = {};
    }
}

void NxapiClient::save_cache() const {
    if (cache_file_.empty()) return;
    try {
        std::filesystem::create_directories(cache_file_.parent_path());
        const Json json(Json::object{
            {"nsoVersion", version_},
            {"nsoVersionExpiresAt", epoch_seconds(version_expiry_)},
            {"rateLimitUntil", epoch_seconds(rate_limit_until_)},
        });
        auto temporary = cache_file_;
        temporary += ".tmp";
        {
            std::ofstream file(temporary, std::ios::binary | std::ios::trunc);
            if (!file) return;
            file << json.dump();
            file.flush();
            if (!file) return;
        }
#ifndef _WIN32
        chmod(temporary.c_str(), 0600);
#endif
        replace_cache_file(temporary, cache_file_);
#ifndef _WIN32
        chmod(cache_file_.c_str(), 0600);
#endif
    } catch (...) {
        // Cache persistence is an optimisation; never break login because disk
        // cache storage is unavailable.
    }
}

void NxapiClient::clear_user_auth() {
    auth_generation_.fetch_add(1);
    std::lock_guard lock(auth_mutex_);
    auth_token_.clear();
    refresh_token_.clear();
    auth_expiry_ = {};
}

void NxapiClient::throw_if_rate_limited() const {
    if (Clock::now() < rate_limit_until_) {
        throw std::runtime_error(
            "nxapi-znca-api Retry-After backoff is active");
    }
}

void NxapiClient::apply_rate_limit_response(const HttpResponse& response) {
    if (response.status == 401) {
        // Do not retry automatically. Forget the rejected in-memory bearer
        // token so the next independent user action can obtain a new one.
        clear_user_auth();
    }

    if (response.status == 406) {
        // The cached Nintendo Switch Online version is explicitly unsupported.
        // Do not retry this request; force the next independent action to fetch
        // /config instead of repeatedly sending a known-bad version.
        version_.clear();
        version_expiry_ = {};
        save_cache();
    }

    if (response.status != 429) return;

    auto delay = std::chrono::duration_cast<std::chrono::seconds>(
        kDefaultRateLimitBackoff);
    if (const auto header = response.headers.find("retry-after");
        header != response.headers.end()) {
        delay = retry_after_delay(header->second);
    }

    rate_limit_until_ = Clock::now() + delay;
    save_cache();
}

std::string NxapiClient::auth_token() {
    // Serialize token acquisition/refresh, but never hold the small state mutex
    // across network I/O. Signing out must be able to invalidate credentials
    // immediately even if nxapi-auth is slow or unavailable.
    std::lock_guard request_lock(auth_request_mutex_);
    const auto generation = auth_generation_.load();
    const auto now = Clock::now();

    std::string refresh_token;
    {
        std::lock_guard state_lock(auth_mutex_);
        if (!auth_token_.empty() && now < auth_expiry_) return auth_token_;
        refresh_token = refresh_token_;
    }

    const auto request_token = [&](const std::string& grant_type,
                                   const std::string& token_to_refresh) {
        std::map<std::string, std::string> fields{
            {"grant_type", grant_type},
            {"client_id", client_id_},
            {"scope", "ca:gf ca:er ca:dr"},
        };
        if (!token_to_refresh.empty()) fields["refresh_token"] = token_to_refresh;
        auto response = http_.post(
            kNxapiAuthUrl,
            HttpClient::form_encode(fields),
            {
                "Accept: application/json",
                std::string("User-Agent: ") + kUserAgent,
            },
            "application/x-www-form-urlencoded");
        // The auth service can rate-limit too. Never retry it automatically;
        // persist any Retry-After so a restart cannot bypass the backoff.
        apply_rate_limit_response(response);
        return response;
    };

    HttpResponse response;
    if (!refresh_token.empty()) {
        response = request_token("refresh_token", refresh_token);
        if (response.status / 100 != 2) {
            std::string error_code;
            try {
                error_code = Json::parse(response.text()).string("error");
            } catch (...) {
            }
            if (error_code != "invalid_grant") {
                throw std::runtime_error(
                    "nxapi-auth refresh failed: " + response.text());
            }
            refresh_token.clear();
            std::lock_guard state_lock(auth_mutex_);
            if (auth_generation_.load() == generation) refresh_token_.clear();
        }
    }

    if (refresh_token.empty() || response.status / 100 != 2) {
        response = request_token("client_credentials", "");
    }
    if (response.status / 100 != 2) {
        throw std::runtime_error(
            "nxapi-auth token request failed: " + response.text());
    }

    const auto json = Json::parse(response.text());
    const auto access_token = json.string("access_token");
    const auto new_refresh_token = json.string("refresh_token", refresh_token);
    if (access_token.empty()) {
        throw std::runtime_error("nxapi-auth response missing access_token");
    }
    const auto expires_in = json.integer("expires_in", 300);

    if (auth_generation_.load() != generation) {
        throw std::runtime_error("nxapi authentication cancelled");
    }
    {
        std::lock_guard state_lock(auth_mutex_);
        if (auth_generation_.load() != generation) {
            throw std::runtime_error("nxapi authentication cancelled");
        }
        auth_token_ = access_token;
        refresh_token_ = new_refresh_token;
        // Public API terms require caching this token for its full validity
        // period. Measure from receipt rather than request start.
        auth_expiry_ = Clock::now() + std::chrono::seconds(
            std::max<std::int64_t>(1, expires_in));
    }
    return access_token;
}

HttpResponse NxapiClient::znca_request(
    const std::string& method,
    const std::string& path,
    const std::string& body,
    const std::string& accept,
    const std::vector<std::string>& extra_headers) {
    // Serialize all requests. The public API requires at most one concurrent
    // automated request, and serialising interactive work as well avoids bursts.
    std::lock_guard lock(request_mutex_);
    throw_if_rate_limited();

    std::vector<std::string> headers{
        "Accept: " + accept,
        std::string("User-Agent: ") + kUserAgent,
    };
    const auto bearer_token = auth_token();
    if (!bearer_token.empty()) {
        headers.push_back("Authorization: Bearer " + bearer_token);
    }
    headers.insert(headers.end(), extra_headers.begin(), extra_headers.end());

    const auto response = http_.request(
        method,
        std::string(kZncaBaseUrl) + path,
        headers,
        std::vector<unsigned char>(body.begin(), body.end()),
        body.empty() ? "" : "application/json",
        30);
    apply_rate_limit_response(response);
    return response;
}

std::string NxapiClient::nso_version() {
    std::lock_guard lock(request_mutex_);
    const auto now = Clock::now();
    if (!version_.empty() && now < version_expiry_) return version_;

    const auto response = znca_request("GET", "/config", "", "application/json");
    if (response.status / 100 == 2) {
        const auto supported_version = Json::parse(response.text()).string("nso_version");
        if (!supported_version.empty()) {
            version_ = supported_version;
            version_expiry_ = now + kVersionCacheLifetime;
            save_cache();
            return version_;
        }
    }

    if (!version_.empty()) {
        version_expiry_ = now + kVersionRetryLifetime;
        save_cache();
        return version_;
    }
    throw std::runtime_error(
        "Could not retrieve supported NSO version from nxapi /config");
}

std::vector<unsigned char> NxapiClient::encrypted_login_body(
    const std::string& id_token,
    const UserProfile& profile) {
    const auto version = nso_version();

    // `/f` output is request-specific (it includes the request timestamp/id and
    // encrypted login body), so it must not be reused as a long-lived token.
    // The reusable credential is the Coral access token returned by Account/Login;
    // CoralClient persists that credential for its server-provided lifetime.
    const Json payload(Json::object{
        {"token", id_token},
        {"hash_method", "1"},
        {"na_id", profile.id},
        {"encrypt_token_request",
         Json::object{
             {"url", kCoralLoginUrl},
             {"parameter",
              Json::object{
                  {"naIdToken", id_token},
                  {"naBirthday", profile.birthday},
                  {"naCountry", profile.country},
                  {"language", profile.language},
                  {"f", ""},
                  {"requestId", ""},
                  {"timestamp", 0},
              }},
         }},
    });

    const auto response = znca_request(
        "POST", "/f", payload.dump(), "application/json", znca_headers(version));
    if (response.status / 100 != 2) {
        throw std::runtime_error("nxapi /f failed: " + response.text());
    }
    const auto encrypted =
        Json::parse(response.text()).string("encrypted_token_request");
    if (encrypted.empty()) {
        throw std::runtime_error("nxapi /f missing encrypted_token_request");
    }
    return base64_decode(encrypted);
}

FAttestation NxapiClient::generate_f(
    int hash_method,
    const std::string& token,
    const std::string& na_id,
    const std::string& coral_user_id) {
    const auto version = nso_version();
    const Json payload(Json::object{
        {"hash_method", std::to_string(hash_method)},
        {"token", token},
        {"na_id", na_id},
        {"coral_user_id", coral_user_id},
    });

    const auto response = znca_request(
        "POST", "/f", payload.dump(), "application/json", znca_headers(version));
    if (response.status / 100 != 2) {
        throw std::runtime_error("nxapi /f failed: " + response.text());
    }
    const auto json = Json::parse(response.text());
    FAttestation attestation;
    attestation.f = json.string("f");
    attestation.request_id = json.string("request_id");
    attestation.timestamp = json.integer("timestamp");
    if (attestation.f.empty() || attestation.request_id.empty() || attestation.timestamp == 0) {
        throw std::runtime_error("nxapi /f returned incomplete attestation: " + response.text());
    }
    return attestation;
}

std::vector<unsigned char> NxapiClient::encrypted_web_service_token_body(
    const std::string& coral_access_token,
    const std::string& na_id,
    const std::string& coral_user_id,
    std::uint64_t game_service_id) {
    const auto version = nso_version();

    const Json payload(Json::object{
        {"token", coral_access_token},
        {"hash_method", "2"},
        {"na_id", na_id},
        {"coral_user_id", coral_user_id},
        {"encrypt_token_request",
         Json::object{
             {"url", "https://api-lp1.znc.srv.nintendo.net/v4/Game/GetWebServiceToken"},
             {"parameter",
              Json::object{
                  {"id", static_cast<double>(game_service_id)},
                  {"registrationToken", ""},
                  {"f", ""},
                  {"requestId", ""},
                  {"timestamp", 0},
              }},
         }},
    });

    const auto response = znca_request(
        "POST", "/f", payload.dump(), "application/json", znca_headers(version));
    if (response.status / 100 != 2) {
        throw std::runtime_error(
            "nxapi /f (hash_method 2) failed: " + response.text());
    }
    const auto encrypted =
        Json::parse(response.text()).string("encrypted_token_request");
    if (encrypted.empty()) {
        throw std::runtime_error(
            "nxapi /f (hash_method 2) missing encrypted_token_request");
    }
    return base64_decode(encrypted);
}

std::vector<unsigned char> NxapiClient::encrypt_request(
    const std::string& url,
    const std::string& coral_token,
    const std::string& json) {
    const auto version = nso_version();
    const Json payload(Json::object{
        {"url", url},
        {"token", coral_token.empty() ? Json(nullptr) : Json(coral_token)},
        {"data", json},
    });
    const auto response = znca_request(
        "POST", "/encrypt-request", payload.dump(), "application/json",
        znca_headers(version));
    if (response.status / 100 != 2) {
        throw std::runtime_error(
            "nxapi /encrypt-request failed: " + response.text());
    }
    const auto encrypted = Json::parse(response.text()).string("data");
    if (encrypted.empty()) {
        throw std::runtime_error("nxapi /encrypt-request missing data");
    }
    return base64_decode(encrypted);
}

std::string NxapiClient::decrypt_response(
    const std::vector<unsigned char>& body) {
    const auto version = nso_version();
    const Json payload(Json::object{{"data", base64_encode(body)}});
    const auto response = znca_request(
        "POST", "/decrypt-response", payload.dump(), "text/plain",
        znca_headers(version));
    if (response.status / 100 != 2) {
        throw std::runtime_error(
            "nxapi /decrypt-response failed: " + response.text());
    }
    return response.text();
}

}  // namespace nso

File: src/nxapi.rs
//! nxapi-znca-api integration used for Nintendo attestation and Coral request crypto.

use crate::http::{DEFAULT_GET_RESPONSE_LIMIT, HttpClient, HttpResponse};
use crate::nintendo_auth::UserProfile;
use crate::util::{base64_decode, base64_standard, json_i64, json_string};
use atomic_write_file::AtomicWriteFile;
use serde_json::json;
use std::fs;
use std::io::Write as _;
use std::path::PathBuf;
use std::sync::Mutex;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

const ZNCA_BASE: &str = "https://nxapi-znca-api.fancy.org.uk/api/znca";
const AUTH_URL: &str = "https://nxapi-auth.fancy.org.uk/api/oauth/token";
const USER_AGENT: &str = "nso-album-sync/2.0.0 (+https://github.com/Dycool/NSO-Album-Sync)";
const CLIENT_VERSION: &str = "w8zSLBsxR7rVoGJA";

#[derive(Debug, Clone, Default)]
pub struct FAttestation {
    f: String,
    request_id: String,
    timestamp: i64,
}

impl FAttestation {
    pub(crate) fn from_parts(f: String, request_id: String, timestamp: i64) -> Self {
        Self { f, request_id, timestamp }
    }

    pub fn f(&self) -> &str {
        &self.f
    }

    pub fn request_id(&self) -> &str {
        &self.request_id
    }

    pub fn timestamp(&self) -> i64 {
        self.timestamp
    }
}

#[derive(Default)]
struct State {
    version: String,
    version_expiry: u64,
    auth_token: String,
    refresh_token: String,
    auth_expiry: u64,
    rate_limit_until: u64,
    generation: u64,
}

pub struct NxapiClient {
    http: HttpClient,
    client_id: String,
    cache_file: PathBuf,
    state: Mutex<State>,
    request_lock: Mutex<()>,
}

impl NxapiClient {
    pub fn new(http: HttpClient, client_id: impl Into<String>, cache_file: PathBuf) -> Self {
        let client = Self {
            http,
            client_id: client_id.into(),
            cache_file,
            state: Mutex::new(State::default()),
            request_lock: Mutex::new(()),
        };
        client.load_cache();
        client
    }

    pub fn clear_user_auth(&self) {
        let mut state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
        state.auth_token.clear();
        state.refresh_token.clear();
        state.auth_expiry = 0;
        state.generation = state.generation.wrapping_add(1);
    }

    pub fn nso_version(&self) -> anyhow::Result<String> {
        let now = now_seconds();
        {
            let state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
            if !state.version.is_empty() && now < state.version_expiry {
                return Ok(state.version.clone());
            }
        }

        match self.znca_request("GET", "/config", "", "application/json", &[]) {
            Ok(response) if response.status() / 100 == 2 => {
                let body: serde_json::Value = serde_json::from_slice(response.body())?;
                let version = json_string(&body, "nso_version");
                anyhow::ensure!(!version.is_empty(), "nxapi config response missing nso_version");
                {
                    let mut state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
                    state.version = version.clone();
                    state.version_expiry = now.saturating_add(6 * 60 * 60);
                }
                self.save_cache();
                Ok(version)
            }
            Ok(response) => self.stale_version_or_error(format!(
                "nxapi config failed (HTTP {})",
                response.status()
            )),
            Err(error) => self.stale_version_or_error(error.to_string()),
        }
    }

    fn stale_version_or_error(&self, message: String) -> anyhow::Result<String> {
        let value = {
            let mut state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
            if state.version.is_empty() {
                None
            } else {
                state.version_expiry = now_seconds().saturating_add(15 * 60);
                Some(state.version.clone())
            }
        };
        if let Some(value) = value {
            self.save_cache();
            Ok(value)
        } else {
            anyhow::bail!(message)
        }
    }

    pub fn generate_f(
        &self,
        hash_method: i32,
        token: &str,
        na_id: &str,
        coral_user_id: &str,
    ) -> anyhow::Result<FAttestation> {
        let _ = self.nso_version()?;
        let body = json!({
            "hash_method": hash_method.to_string(),
            "token": token,
            "na_id": na_id,
            "coral_user_id": coral_user_id,
        });
        let response = self.znca_request(
            "POST",
            "/f",
            &serde_json::to_string(&body)?,
            "application/json",
            &[],
        )?;
        anyhow::ensure!(
            response.status() / 100 == 2,
            "nxapi f request failed (HTTP {})",
            response.status()
        );
        let body: serde_json::Value = serde_json::from_slice(response.body())?;
        let result = body.get("result").unwrap_or(&body);
        let output = FAttestation::from_parts(
            json_string(result, "f"),
            json_string(result, "request_id"),
            json_i64(result, "timestamp", 0),
        );
        anyhow::ensure!(
            !output.f.is_empty() && !output.request_id.is_empty() && output.timestamp > 0,
            "nxapi returned incomplete f attestation"
        );
        Ok(output)
    }

    pub fn encrypted_login_body(
        &self,
        id_token: &str,
        profile: &UserProfile,
    ) -> anyhow::Result<Vec<u8>> {
        let attestation = self.generate_f(1, id_token, profile.id(), "")?;
        let body = json!({"parameter": {
            "naIdToken": id_token,
            "birthday": profile.birthday(),
            "country": profile.country(),
            "language": profile.language(),
            "f": attestation.f(),
            "requestId": attestation.request_id(),
            "timestamp": attestation.timestamp()
        }});
        self.encrypt_request(
            "https://api-lp1.znc.srv.nintendo.net/v4/Account/Login",
            "",
            &serde_json::to_string(&body)?,
        )
    }

    pub fn encrypted_web_service_token_body(
        &self,
        coral_access_token: &str,
        na_id: &str,
        coral_user_id: &str,
        game_service_id: u64,
    ) -> anyhow::Result<Vec<u8>> {
        let attestation = self.generate_f(2, coral_access_token, na_id, coral_user_id)?;
        let body = json!({"parameter": {
            "id": game_service_id,
            "f": attestation.f(),
            "requestId": attestation.request_id(),
            "timestamp": attestation.timestamp()
        }});
        self.encrypt_request(
            "https://api-lp1.znc.srv.nintendo.net/v4/Game/GetWebServiceToken",
            coral_access_token,
            &serde_json::to_string(&body)?,
        )
    }

    pub fn encrypt_request(
        &self,
        url: &str,
        coral_token: &str,
        plaintext_json: &str,
    ) -> anyhow::Result<Vec<u8>> {
        let _ = self.nso_version()?;
        let body = json!({
            "url": url,
            "token": if coral_token.is_empty() {
                serde_json::Value::Null
            } else {
                serde_json::Value::String(coral_token.to_owned())
            },
            "data": serde_json::from_str::<serde_json::Value>(plaintext_json)
                .unwrap_or_else(|_| serde_json::Value::String(plaintext_json.to_owned()))
        });
        let response = self.znca_request(
            "POST",
            "/encrypt-request",
            &serde_json::to_string(&body)?,
            "application/json",
            &[],
        )?;
        anyhow::ensure!(
            response.status() / 100 == 2,
            "nxapi encrypt request failed (HTTP {})",
            response.status()
        );
        let body: serde_json::Value = serde_json::from_slice(response.body())?;
        let encoded = json_string(&body, "data");
        anyhow::ensure!(!encoded.is_empty(), "nxapi encrypt response missing data");
        base64_decode(&encoded)
    }

    pub fn decrypt_response(&self, encrypted: &[u8]) -> anyhow::Result<String> {
        let _ = self.nso_version()?;
        let body = json!({"data": base64_standard(encrypted)});
        let response = self.znca_request(
            "POST",
            "/decrypt-response",
            &serde_json::to_string(&body)?,
            "text/plain,application/json",
            &[],
        )?;
        anyhow::ensure!(
            response.status() / 100 == 2,
            "nxapi decrypt response failed (HTTP {})",
            response.status()
        );
        let text = response.text();
        if let Ok(body) = serde_json::from_str::<serde_json::Value>(&text) {
            let data = json_string(&body, "data");
            if !data.is_empty() {
                return Ok(data);
            }
        }
        Ok(text)
    }

    fn znca_request(
        &self,
        method: &str,
        path: &str,
        body: &str,
        accept: &str,
        extra_headers: &[String],
    ) -> anyhow::Result<HttpResponse> {
        let _request_guard = self.request_lock.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
        self.throw_if_rate_limited()?;
        let token = self.auth_token()?;
        let version = self
            .state
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner())
            .version
            .clone();
        let mut headers = vec![
            format!("Authorization: Bearer {token}"),
            format!("User-Agent: {USER_AGENT}"),
            format!("Accept: {accept}"),
            "X-znca-Platform: Android".to_owned(),
            format!("X-znca-Client-Version: {CLIENT_VERSION}"),
        ];
        if !version.is_empty() {
            headers.push(format!("X-znca-Version: {version}"));
        }
        headers.extend_from_slice(extra_headers);
        let url = format!("{ZNCA_BASE}{path}");
        let response = if method == "GET" {
            self.http
                .get(&url, &headers, 30, DEFAULT_GET_RESPONSE_LIMIT)?
        } else {
            self.http.post_text(
                &url,
                body,
                &headers,
                "application/json",
                30,
                DEFAULT_GET_RESPONSE_LIMIT,
            )?
        };
        self.apply_rate_limit_response(&response);
        if response.status() == 401 {
            let mut state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
            state.auth_token.clear();
            state.auth_expiry = 0;
        }
        if response.status() == 406 {
            {
                let mut state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
                state.version.clear();
                state.version_expiry = 0;
            }
            self.save_cache();
        }
        Ok(response)
    }

    fn auth_token(&self) -> anyhow::Result<String> {
        let now = now_seconds();
        let (refresh, generation) = {
            let state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
            if !state.auth_token.is_empty() && now < state.auth_expiry {
                return Ok(state.auth_token.clone());
            }
            (state.refresh_token.clone(), state.generation)
        };
        anyhow::ensure!(
            !self.client_id.trim().is_empty(),
            "nxapi auth client id is missing"
        );

        let mut response = None;
        let mut previous_refresh = refresh.clone();
        if !refresh.is_empty() {
            let fields = [
                ("grant_type", "refresh_token"),
                ("client_id", self.client_id.as_str()),
                ("scope", "ca:gf ca:er ca:dr"),
                ("refresh_token", refresh.as_str()),
            ];
            let refresh_response = self.http.post_form(
                AUTH_URL,
                &fields,
                &[format!("User-Agent: {USER_AGENT}"), "Accept: application/json".to_owned()],
                30,
            )?;
            self.apply_rate_limit_response(&refresh_response);
            if refresh_response.status() / 100 == 2 {
                response = Some(refresh_response);
            } else {
                let error_code = serde_json::from_slice::<serde_json::Value>(refresh_response.body())
                    .ok()
                    .map(|value| json_string(&value, "error"))
                    .unwrap_or_default();
                anyhow::ensure!(
                    error_code == "invalid_grant",
                    "nxapi auth refresh failed (HTTP {}): {}",
                    refresh_response.status(),
                    refresh_response.text()
                );
                previous_refresh.clear();
                let mut state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
                if state.generation == generation {
                    state.refresh_token.clear();
                }
            }
        }

        let response = match response {
            Some(response) => response,
            None => {
                let fields = [
                    ("grant_type", "client_credentials"),
                    ("client_id", self.client_id.as_str()),
                    ("scope", "ca:gf ca:er ca:dr"),
                ];
                let response = self.http.post_form(
                    AUTH_URL,
                    &fields,
                    &[format!("User-Agent: {USER_AGENT}"), "Accept: application/json".to_owned()],
                    30,
                )?;
                self.apply_rate_limit_response(&response);
                response
            }
        };
        anyhow::ensure!(
            response.status() / 100 == 2,
            "nxapi auth failed (HTTP {}): {}",
            response.status(),
            response.text()
        );
        let body: serde_json::Value = serde_json::from_slice(response.body())?;
        let access = json_string(&body, "access_token");
        anyhow::ensure!(!access.is_empty(), "nxapi auth response missing access_token");
        let expires = json_i64(&body, "expires_in", 3600).max(1) as u64;
        let next_refresh = body
            .get("refresh_token")
            .and_then(serde_json::Value::as_str)
            .unwrap_or(&previous_refresh)
            .to_owned();

        let mut state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
        anyhow::ensure!(
            state.generation == generation,
            "nxapi user auth was cleared during token refresh"
        );
        state.auth_token = access.clone();
        state.refresh_token = next_refresh;
        state.auth_expiry = now_seconds().saturating_add(expires);
        Ok(access)
    }

    fn throw_if_rate_limited(&self) -> anyhow::Result<()> {
        let until = self
            .state
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner())
            .rate_limit_until;
        anyhow::ensure!(
            now_seconds() >= until,
            "nxapi request is locally rate-limited until unix time {until}"
        );
        Ok(())
    }

    fn apply_rate_limit_response(&self, response: &HttpResponse) {
        if response.status() != 429 {
            return;
        }
        let retry = response
            .header("retry-after")
            .and_then(|value| value.trim().parse::<u64>().ok())
            .unwrap_or(15 * 60)
            .clamp(1, 60 * 60);
        self.state
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner())
            .rate_limit_until = now_seconds().saturating_add(retry);
        self.save_cache();
    }

    fn load_cache(&self) {
        let Ok(bytes) = fs::read(&self.cache_file) else {
            return;
        };
        let Ok(body) = serde_json::from_slice::<serde_json::Value>(&bytes) else {
            return;
        };
        let mut state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
        state.version = json_string(&body, "nsoVersion");
        state.version_expiry = body
            .get("versionExpiry")
            .and_then(serde_json::Value::as_u64)
            .unwrap_or(0);
        state.rate_limit_until = body
            .get("rateLimitUntil")
            .and_then(serde_json::Value::as_u64)
            .unwrap_or(0);
    }

    fn save_cache(&self) {
        let snapshot = {
            let state = self.state.lock().unwrap_or_else(|poisoned| poisoned.into_inner());
            json!({
                "nsoVersion": state.version,
                "versionExpiry": state.version_expiry,
                "rateLimitUntil": state.rate_limit_until
            })
        };
        let Some(parent) = self.cache_file.parent() else {
            return;
        };
        if fs::create_dir_all(parent).is_err() {
            return;
        }
        let Ok(bytes) = serde_json::to_vec_pretty(&snapshot) else {
            return;
        };
        let Ok(mut file) = AtomicWriteFile::open(&self.cache_file) else {
            return;
        };
        let commit_ok = file.write_all(&bytes).is_ok()
            && file.flush().is_ok()
            && file.commit().is_ok();
        #[cfg(unix)]
        if commit_ok {
            use std::os::unix::fs::PermissionsExt as _;
            let _ = fs::set_permissions(&self.cache_file, fs::Permissions::from_mode(0o600));
        }
        #[cfg(not(unix))]
        let _ = commit_ok;
    }
}

fn now_seconds() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or(Duration::ZERO)
        .as_secs()
}
